storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+				attachArg.Provider,
+			)
+			continue
+		}
+
+		attachment, err := s.openstackClient.AttachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			"", // TODO(katco): Fill this in when attachment support is added to parameters.
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		volAttachments[argIdx] = attachment
+	}
+
+	return volAttachments, nil
+}
+
+// DetachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DetachVolumes(params []storage.VolumeAttachmentParams) error {
+	panic("not implemented")
+}
+
+type OpenstackAdapter interface {
+	GetVolumesSimple() ([]storage.Volume, error)
+	DeleteVolume(string) error
+	CreateVolume(size uint64, name string) (storage.Volume, error)
+	AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error)
+}
+
+func newGooseAdapter(cfg *storage.Config) (OpenstackAdapter, error) {
+
+	region, _ := cfg.ValueString("region")
+
+	authClient := client.NewClient(&identity.Credentials{
+		URL:        cfg.Attrs()[openstackCfgKeyUrl].(string),
+		User:       cfg.Attrs()[openstackCfgKeyUser].(string),
+		Secrets:    cfg.Attrs()[openstackCfgKeySecrets].(string),
+		Region:     region,
+		TenantName: cfg.Attrs()[openstackCfgKeyTenantName].(string),
+	}, identity.AuthUserPass, nil)
+
+	endpoint := authClient.EndpointsForRegion(region)["volume"]
+	endpointUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, errors.Annotate(err, "error parsing endpoint")
+	}
+	cinderClient := cinder.Basic(endpointUrl, authClient.TenantId(), authClient.Token)
+
+	novaClient := nova.New(authClient)
+
+	return &gooseAdapter{cinderClient, novaClient}, nil
+}
+
+type gooseAdapter struct {
+	cinder *cinder.Client
+	nova   *nova.Client
+}
+
+func (ga *gooseAdapter) CreateVolume(size uint64, name string) (storage.Volume, error) {
+
+	resp, err := ga.cinder.CreateVolume(cinder.CreateVolumeVolumeParams{
+		// NOTE: The Cinder documentation incorrectly states the
+		// size parameter is in GB. It is actually GiB.
+		Size: int(math.Ceil(float64(size / 1024))),
+		Name: name,
+	})
+	if err != nil {
+		return storage.Volume{}, err
+	}
+	return cinderToJujuVolume(&resp.Volume), err
+}
+
+func (ga *gooseAdapter) DeleteVolume(volId string) error {
+	return ga.cinder.DeleteVolume(volId)
+}
+
+func (ga *gooseAdapter) GetVolumesSimple() ([]storage.Volume, error) {
+
+	resp, err := ga.cinder.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	vols := make([]storage.Volume, len(resp.Volumes))
+	for idx, vol := range resp.Volumes {
+		vols[idx] = cinderToJujuVolume(&vol)
+	}
+
+	return vols, nil
+}
+
+func (ga *gooseAdapter) AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error) {
+	attachment, err := ga.nova.AttachVolume(serverId, volumeId, mountPoint)
+	return storage.VolumeAttachment{
+		Volume:     names.NewVolumeTag(attachment.VolumeId),
+		Machine:    names.NewMachineTag(attachment.ServerId),
+		DeviceName: attachment.Device,
```

Comment:
One thing to bear in mind: Juju's DeviceName is the bare device name, not the path. e.g. "sda" and not "/dev/sda". I'm not sure what it is in Nova/Cinder, but EC2/EBS returns the path. What's more, in EC2 you specify, say, "/dev/sda" but that actually gets translated to "/dev/xvda" on the machine that's created (in recent Linux kernels). So there's some translation required for EC2 - not sure what the case is here.